stdlib/Printf/src/Printf.jl

```diff
@@ -2,1297 +2,770 @@
-    else
-        push!(blk.args, :(unsafe_write(out, pointer(buf), len)))
-        push!(blk.args, :(while pt >= (len+=1) print(out,'0') end))
-        '#' in flags && push!(blk.args, :(print(out, '.')))
-    end
-    # print space padding
-    if padding !== nothing && '-' in flags
-        push!(blk.args, pad(width-1, padding, ' '))
-    end
-    # return arg, expr
-    :(($x)::Real), ex
-end
+For arbitrary precision numerics, you might extend the method like:
 
-function gen_e(flags::String, width::Int, precision::Int, c::Char, inside_g::Bool=false)
-    # print float in scientific form:
-    #  [e]: use 'e' to introduce exponent
-    #  [E]: use 'E' to introduce exponent
-    #
-    # flags:
-    #  (#): always print a decimal point
-    #  (0): pad left with zeros
-    #  (-): left justify
-    #  ( ): precede non-negative values with " "
-    #  (+): precede non-negative values with "+"
-    #
-    x, ex, blk = if inside_g
-        @gensym x
-        blk = Expr(:block)
-        x, blk, blk
-    else
-        special_handler(flags,width)
-    end
-    # interpret the number
-    if precision < 0; precision = 6; end
-    ndigits = min(precision+1,length(Grisu.getbuf())-1)
-    push!(blk.args, :((do_out, args) = ini_dec(out,$x,$ndigits, $flags, $width, $precision, $c, buf)))
-    push!(blk.args, :(digits = buf))
-    ifblk = Expr(:if, :do_out, Expr(:block))
-    push!(blk.args, ifblk)
-    blk = ifblk.args[2]
-    push!(blk.args, :((len, pt, neg) = args))
-    push!(blk.args, :(exp = pt-1))
-    expmark = isuppercase(c) ? "E" : "e"
-    if precision==0 && '#' in flags
-        expmark = string(".",expmark)
-    end
-    # calculate padding
-    padding = nothing
-    width -= precision+length(expmark)+(precision>0)+4
-    # 4 = leading + expsign + 2 exp digits
-    if '+' in flags || ' ' in flags
-        width -= 1 # for the sign indicator
-        if width > 0
-            padding = quote
-                padn=$width
-                if (exp<=-100)|(100<=exp)
-                    if isa($x,SmallNumber)
-                        padn -= 1
-                    else
-                        padn -= Base.ndigits0z(exp) - 2
+```julia
+Printf.tofloat(x::MyArbitraryPrecisionType) = BigFloat(x)
+```
+"""
+tofloat(x) = Float64(x)
+tofloat(x::Base.IEEEFloat) = x
+tofloat(x::BigFloat) = x
+
+@inline function fmt(buf, pos, arg, spec::Spec{T}) where {T <: Floats}
+    leftalign, plus, space, zero, hash, width, prec =
+        spec.leftalign, spec.plus, spec.space, spec.zero, spec.hash, spec.width, spec.precision
+    x = tofloat(arg)
+    if x isa BigFloat && isfinite(x)
+        ptr = pointer(buf, pos)
+        newpos = @ccall "libmpfr".mpfr_snprintf(ptr::Ptr{UInt8}, (length(buf) - pos + 1)::Csize_t, string(spec; modifier="R")::Ptr{UInt8}; arg::Ref{BigFloat})::Cint
+        newpos > 0 || error("invalid printf formatting for BigFloat")
+        return pos + newpos
+    elseif x isa BigFloat
+        x = Float64(x)
+    end
```

Comment:
This needs to not be hardcoded here (especially since Keno has just proposed removing BigFloat)

Reply:
What exactly shouldn't be hard-coded? The library name? If we move BigFloat to stdlib, wouldn't Printf just take a dependency on it?

Reply:
why not make it a function interface which we can intercept? e.g. so that we could extend it for Quadmath.jl or DecFP.jl

Reply:
There shouldn't be a list of known types in this module—that's why Julia has dispatch

Reply:
Please see previous comments [here](https://github.com/JuliaLang/julia/pull/32859#issuecomment-521512416) and [here](https://github.com/JuliaLang/julia/pull/32859#issuecomment-521839355) for the compat/intercept story. In short, there are already ways to overload/provide for custom types to hook in to Printf at various levels. And we're at least providing solutions to all known users/overloaders of current Printf code, in most cases allowing for a much simpler integration.

Reply:
Yes, so this should be dog-fooding whatever capability it has. I think possibly there's a mismatch here on implementation vs policy, where this function is confused as to which it is doing?

I've added the needs-docs tag, since that comment (on how to integrate this with a custom type) would be good to have in the PR.

Reply:
It is? Up on line 371, it calls `x = tofloat(arg)`, which is one of the documented ways to hook a custom type into printf integration.

Reply:
If true, there wouldn't be a branch here naming a random external type as `x isa BigFloat`. Why is that present?

Reply:
Maybe look more closely at the code? The logic is as follows:
* Call `x = tofloat(arg)`, which is documented such that overloaders need to return one of `Float16`, `Float32`, `Float64`, or `BigFloat`. 
* If `tofloat` returned a `BigFloat` (as `BigFloat` does itself), and is finite, we call libmpfr to format it
* If the BigFloat is not finite, (i.e. `NaN`, `Inf`, or `-Inf`), we convert to `Float64` to then be formatted

Does that make sense? I might be missing a valid concern here, so please explain in more detail if so.